### ISSUE TITLE
T | (() => T) 

### ISSUE THREAD

**Issue by dhmk083** [microsoft/TypeScript#37663]
**TypeScript Version:**  3.9.0 (Nightly)

**Search Terms:** `T | (() => T)`, `T & Function`

**Code**

```ts
type Initializer<T> = T | (() => T)
// type Initializer<T> = T extends any ? (T | (() => T)) : never

function correct<T>(arg: Initializer<T>) {
    return typeof arg === 'function' ? arg() : arg // error
}
```

Line 2 provides a workaround for this.
More info on [stackoverflow](https://stackoverflow.com/questions/60898079/typescript-type-t-or-function-t-usage).

**Expected behavior:** no errors

**Actual behavior:** `This expression is not callable.
  Not all constituents of type '(() => T) | (T & Function)' are callable.
    Type 'T & Function' has no call signatures.`

**Playground Link:** [here](https://www.typescriptlang.org/play/?ts=3.9.0-dev.20200327&ssl=1&ssc=1&pln=6&pc=2#code/C4TwDgpgBAkgdgS2AghgGwQLwgJwDwAqAfFALxQFQA+UAFLQJRkkEMBQA9B1KJLIsnRZchEuUoQAHsAhwAJgGcoKOCCgB+OpRr0mpFgyYAuKHAgA3XGzYAzAK5wAxsgD2cKI5c4cEZ6NooOADmJvBIqBjY+MRMAN5sUIlQPsB2OO68EC42ysFkpOQA5PZOrnCFGrlBjFAmgUFQXFC4OF5sAL5AA).

**Related Issues:** none

**Comment by DanielRosenwasser** [comment]
This check isn't safe if `T` is instantiated with something like `(x: string) => number`.

**Comment by bela53** [comment]
@DanielRosenwasser Why does the following alternative work then?
```
type Initializer<T> = T extends any ? (T | (() => T)) : never

function correct<T>(arg: Initializer<T>): T {
    return typeof arg === 'function' ? arg() : arg // arg is callable in the true branch
}
```
[Playground](https://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=20&pc=71#code/C4TwDgpgBAkgdgS2AghgGwQLwgJwDwAqAfFALxQFQQAewEcAJgM5QpwhQD8UAFJQD68eASjIkCw0QC4ocCADdcAWABQqgGYBXOAGNkAezhQd+nDgh7CRHihwBzGfCSoM2fMWEzKAb1VR-UObAmjhGoJD66qz2ZKTkAORaugZw8VzRdiJQMrZ2UAD0+VAA7qYA1kx+AYUZUAgsTsjoWLhWUABkUABi2noIhnVhABbQwDia0ABGOGw6Q6oAvqqqJnBMwIEAjGTGpuZ6PABMojWr61syhyuG5ziHOyZmFsA8WaQkxwVFZxt3MnCaAC2k1wQA)

`arg` becomes `Initializer<T> & Function` in the true branch and *is* callable. This example seems quite similar to me with the only difference, that `Initializer<T>` from OP is a union function type, above not. We could ask the same question here: is the check for `T` safe? 

In addition, I expected `Function` to be a very generous function declaration. Today you already can do invocations with all sorts of arguments:
```
declare const foo: Function
// all compile
foo()
foo(3)
foo("bar",3, true)
```
So I would expect and wish that behavior at least gets consistent. Either for both examples, a) an error should be emitted or b) both should compile to be in line with the current `Function` type behavior.

Appreciate your comment, thanks!

**Comment by zerkms** [comment]
@bela53 

```typescript
const r2 = correct<(_: string) => number>((v: string) => v.toString().length)
```

this code passes type validation but blows in runtime.

**Comment by bela53** [comment]
@zerkms Good catch. You could fix that error by having a more narrow constraint for `Initializer<T>`:
```
type Initializer<T> = T extends () => any ? (T | (() => T)) : never
const r2 = correct<() => number>(() => "foo".length)
```
Though this isn't directly related to the main issue. `typeof arg === 'function'` is only able to narrow to a `Function` type in the true branch. And in principle, you can invoke `Function` like you want. Question remains, if both alternatives above should work or not.

**Comment by bela53** [comment]
@zerkms Good catch. You could fix that error by choosing a more narrow type relationship test in `Initializer<T>` conditional type:
```
type Initializer<T> = T extends () => any ? (T | (() => T)) :
    T extends Function ? never : (T | (() => T))
const r3 = correct<(_: string) => number>((v: string) => v.toString().length) // error
``` 
However, this isn't directly related to the main issue. `typeof arg === 'function'`  will only be able to narrow to `Function` in the true branch, and in principle you can invoke `Function` with any arguments.

The question remains, if both `Initializer<T> & Function` and `T & Function` are callable or not (only one or the other would be inconsistent in my view).

**Comment by DanielRosenwasser** [comment]
*that* seems like a bug

**Comment by falsandtru** [comment]
Simplified.

```ts
function f<T>(arg: T & Function) {
    typeof arg === 'function' && arg(); // `T & Function` is callable.
}
function g<T>(arg: T | (() => T)) {
    typeof arg === 'function' && arg(); // Type 'T & Function' has no call signatures.(2349)
}
```

**Comment by dhmk083** [comment]
@bela53 

```ts
type Initializer<T> = T extends Function ? never : (T | (() => T))
```
Seems to be working too and is shorter. I think I will use this typing.

By the way, 
```ts
type Initializer<T> = T | (() => T)
// -or-
type Initializer<T> = T extends any ? (T | (() => T)) : never

const isFunction = (arg: any): arg is Function => typeof arg === 'function'

function correct_2<T>(arg: Initializer<T>): T {
    return isFunction(arg) ? arg() : arg
}

const x = correct_2<() => number>(() => 5)
// x: () => number
```
This compiles, but produces a mismatch between type and runtime value.

**Comment by dhmk083** [comment]
Unfortunately, this typing doesn't always work. 
I've found some inconsistencies in its behavior:
```ts
// type Initializer<T> = T | (() => T)
// -or-
type Initializer<T> = T extends Function ? never : T | (() => T)

function f_1<T>(arg: Initializer<T>) {
    return typeof arg === 'function' ? arg() : arg
}

function f_2<T extends number>(arg: Initializer<T>) {
    return typeof arg === 'function' ? arg() : arg
}

function f_3<T>(arg: Initializer<Partial<T>>) {
    return typeof arg === 'function' ? arg() : arg
}

function f_4<T>(arg: Initializer<Record<'a', T>>) {
    return typeof arg === 'function' ? arg() : arg
}
```

[Playground](https://www.typescriptlang.org/play/?ts=3.9.0-dev.20200328&ssl=19&ssc=2&pln=1&pc=1#code/PTAEBcE8AcFNQJIDsCW4UEMA2KBesAnAHgBUA+UAXlBNAB9QAKRgSiopJYCgRQBaAPYE+XKHESp02PIVIVqtWAA9wsJABMAzqABiAVyQBjdAKSgA-KCSwAboVAAuGvSat2NblwBmB4ylOgXgD6AIxyjBgEAOZOyGiYOPjE5GwA3lygmaAEsOB6BGZisAJeoJFRVJTUAOQ+RiZI1RZl0W5O5VwAvlzevg2BQQBMpKDKqhraSHoAtgBGhGQR0bGSCTLJZGkZWTl5BRAwxaXllTV1fqZNluVtLVFdPef9wQDM4eUr8dJJRAAKkVIsHJNqB0llsrl8oVDiU7qdQLU+v5Gs0bmx2tEHr16siBgAWd7LCRfRKyABKsEMQnURGqGGqABoaGQQWCdpD9kVYScqmckZdUa10XcukA)

If you try with line 1 typing, it will fail f_1 case as expected, but it will also work for cases 2,3,4.
If you try with line 3, it will fail for cases 2 and 3.
I don't get why it doesn't work for Partial<T>, but works for Record<K, T>. I think they should either all work or all fail.

**Comment by bela53** [comment]
@falsandtru well summarized the issue. I would add one case:
```
// if you set a constraint for T, it works again (?)
function g2<T extends number>(arg: T | (() => T)) {
    typeof arg === 'function' && arg(); // () => T is callable.
}
```

[Plagyround with all  three examples](https://www.typescriptlang.org/play/?ts=3.9.0-dev.20200328&ssl=12&ssc=2&pln=9&pc=1#code/GYVwdgxgLglg9mABMAPAFQHwAoCGAnAcwC5E1EAyRAMXGnjAEpEBvAKEQ8SgE8AHAUzjBE+AogC8kxAHJQkWAmkVKorAwDciAPRbEAAzKUa8+nsQwAzogg4ANrZwAjW-wB0rAL6tWcugkQE6NiiJGQAPohYahIYpAxMbJxcfILCohJSsrQKYErkKoRqmjqkKTKG1Nn0SgAWOFZgcNZ2togWMARgOFAgePwWrlgATADMACwAnAye3iUwwtxwICIAJisi1ggWUHg4MGBQXE1oADTmhwDucHgA1lY4BHtIWAD80745AUPoiPwAHlB+GAVg0QABbRz8PDBQihRARKJMcSxNDxFjsTg8ARCESEDLiGQfarKXEEIraXTRZGkcxWGz2JwudweIA)

@dhmk083 If I am not mistaken, `f_2` and `f_3` seem to [describe the same problem in the end](https://www.typescriptlang.org/play/?ts=3.9.0-dev.20200328#code/PTAEEMCcHNQSwM6gCqgD6gBSYJSgLwB8KOAUAGYCuAdgMYAucA9taOQPoCMAPMoZlGgAuFKACmAD3pjqAEyQAxGg2asA-KGpiAbmMigRqDNjxESeAN6lQN0JDH1KkVvQCeABzFNyEGAXz4oADkVHSMLEGgGoK4Br7QpAC+pKQg8fBImKiS0nJI1JQAtgBGenjGsWZZ4lIy8ppFpZBkoSosbOwATLw1ufUFJXr8goa9dYrK4eqaOnpxRliVxMg4lta29o7OoG6e3ukBgSGTqpHRMLEigkkpaYIZoAAKUIzgADa8xBWmxM+Qrx8+BQTu0OABmT4CGCjHLjUBKMKqKIzXT6ER-AGfdCLH5PF5wd6fVagKy2OwOJwuDxeHz3Q7BVpTM7xS7xG6pMD3RCgABKYloTEgsm4ACJwCKADQoL44gjEPkCoWi8VSvg4KUAdzg9AAFhBQAK6JsxLz+YLZKARRrIOB3J5ICKdtTgYjQewACyQkamxXCoLgIKq4iwvLwkHTLSouIK83cf2B6XodY2Exyn2x+NB4mkjYU7a7GkHAIM8PMmJ4K4wdmkIA) (after the conditional type is resolved)

**Comment by dhmk083** [comment]
@bela53 
In your last example `T extends number` means that T is definitely not a `Function`, so it works. If you replace it with `T extends object` constraint, it will produce an error, because `Function` extends `object`, too.

Here is another simplified case:
```ts
// OK
function g<T>(arg: T extends Function ? never : (T | (() => T))) {
    typeof arg === 'function' && arg(); // () => T is callable.
}

// error
function g2<T extends object>(arg: T extends Function ? never : (T | (() => T))) {
    typeof arg === 'function' && arg(); // () => T is callable.
}
```

Adding a constraint breaks things.

**Comment by jack-williams** [comment]
Is this related #27422?

**Comment by Svish** [comment]
Ran into this while trying to use the [`SetStateAction<T>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L873) type in `react`:

```ts
type SetStateAction<S> = S | ((prevState: S) => S);
```

If using `typeof foobar === 'function'` _isn't_ the correct way to check whether we have a value, or a callable initializer function, what is?

**Comment by MikeYermolayev** [comment]
FYI: `instanceof` workaround doesn't work for values in generic Record

```
type Schema = {
    id: string;
    type: string;
}

type Config = {
    schema: Schema | (() => Schema)
}

function f<T extends Record<string, Config>>(config: T, key: string) {
    if (config[key] instanceof Function) {
        config[key]() // This expression is not callable
    }
}
```

[Playground](https://www.typescriptlang.org/play?ts=4.2.3#code/C4TwDgpgBAygxgCwgWwIZQLxQN4CgoFQCWAJgFxQDOwATkQHYDmA3PoaJBdXU6wL65cHaAGEA9vQBmRRphxsClRClQV4SNFAA+UABS6AlJgB8sZWgO4BuSQFd6cYEQlRJAHgAqUCAA9gEehJKKAAlCDgxGhI3bgZGABoocSkZY2NdCJTGCg9EgGsIEC5aOKM8QmJJPUzpRgBtApAAXWJ6alQHCDEqgDF7R2d6MoUKqBqZBsKmwxGBPiA)

> Found that this reveals the issue:
> 
> ```js
> function a<T>(b: T | (() => T)) {
>   return typeof b === 'function' ? b() : b
> }
> ```
> 
> But this does not (and compiles file):
> 
> ```ts
> function a<T>(b: T | (() => T)) {
>   return b instanceof Function ? b() : b
> }
> ```
> 
> Kind of odd, because `typeof` checks seem to work normally. But this is a reasonable workaround for anyone who bumps into it. HT to @pbteja1998 :)

**Comment by jsejcksn** [comment]
Be careful with using `x instanceof Function` (as [suggested](https://github.com/microsoft/TypeScript/issues/37663#issuecomment-759728342) by @kentcdodds), as it won't be reliable in cases where the value comes from a different context (e.g. iframe/window/[realm](https://github.com/tc39/proposal-realms)). Objects (including functions) from other contexts [will have a different Function constructor in their prototype chain](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/instanceof#:~:text=they%20have%20different%20built-ins) than the one being used to perform the check.

**Comment by kentcdodds** [comment]
True, I should have mentioned that (I've definitely bumped into this issue in the past and it was *very* confusing at the time 😅). I'd love a solution rather than a workaround :)

**Comment by jsejcksn** [comment]
Here's one that feels a little verbose, but seems to work:

```ts
type Fn = (...args: unknown[]) => unknown;

function getValue <
  Provided,
  T = Provided extends Fn ? ReturnType<Provided> : Provided,
>(valueOrFn: Provided): T {
  return typeof valueOrFn === 'function' ? valueOrFn() : valueOrFn;
}
```

[Playground](https://www.typescriptlang.org/play?strict=true#code/C4TwDgpgBAYgdlAvFAFAOgwQwE4HMDOAXFAK5wDWcA9gO5wDaAugJRIB8pF1dA3AFB8AZmQDGwAJZUEuCMABqmADYloAHj5QoABWxUAbuIAmEQwBoNUACpJtug8cNQIAD2AQ4h-LAQB+KACVZEmw4S3AIVR19IxMOYij7E3M2FD0lFQB5bHh4uxjDZmJrAG8LbCCQqFBIKkEoNOUILPgkRGQAcmE4MUk4dqg-BszsuBRWYiGmkf4AXwFjEUUcaBEpfGAoHFxiMfZbKgBbcXwIwNXsQ1V17HE4XFMoOBIDgCMIbDY2flW4dahy-AkRQbZCYGiYcQbGTydIQFBbZj8PguMBUbAbYozHhQAD0OKgAElcNRylUABbHbEAOQgJhMUEEaKsABkAIJAA)

**Comment by smac89** [comment]
So basically at this point, the only way to make this work is to use `instanceof`??

You know, I could have sworn this method of using `typeof foo === 'function'`, used to work! In fact I just started noticing it in VSCode today.

Was this a regression?

**Comment by rpearce** [comment]
This is the only reasonable way I could find around this without having to cast all the things `as <shape of function>`:

```typescript
const isFunction = (x: unknown): x is Function => typeof x === 'function'
```

[Here's an example of it in action with a nice `cond` function I've been working on](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgMIHsQBNkG8BQyyAPAIIA0yAQgHwAUYArgA4A2EAzlegB4BcyANqEiyUsgA+yOnQBucVgNIBKZAF4ayAEbp07OCGXkRRKpOlyFS1RupGRAXUEPlAuljhg419ZvFSqfABffHwETA4wZHDsAQxsdWQmNk5uHl9kDy8MgiIYyKEAOmLk9g4HRNLU3kkpZ1CiGHQoaXyowRhIyg5sCvQYJJYy1VzRYAG6Tqi1GczPOFrkYA4AMUYQBDBgTEnI1QAyfeQp93nlEZNRKAgmKBAl1fXN7ZA6HqxVAH5kd9OvVQE70uISIIRE11u9yycGCoQA9HDkAALCCsZjQDhhCJRZZrDZbTCJOj8ZDrADWIHQAHdDAJ0stkHjnoTbGAAJ7o-rIdIzNTIADkMCeBJA-KxIAKABUiTZNGAoIwIPgEcgIAkUWiMaF8noIIVWOgAOZ0EQxLB0YSiISC3T8yg8zQAAypcA4x103wAJLgeEFHQ5jFbBA7uepZvytHAoHaBS63ZHowHLoJ+VADFh0ABbGP8ylgJGgQ38pNByWUOiy2Ou5CUkAQYuB5AuOg29D8+ymzDmy2iFNNNv2jLO6v9r0+v0l3shnnhhM5uPaKMN5Op9NZnN5gsgIuToiCMvSSv8he1+uT5sRpcdvJdi0r-sxkPDt2j5De33+xt76dhvmX6OUMe1Zzru1pptg66AZuhbLqWgEnpgZ6NhekYAF7tvgyj4EAA)

**Comment by Elia-Darwish** [comment]
I've been using this type guard recently, and it also seems to solve the problem. It would be nice to have a fix, though!

```ts
declare type AnyFunction = (...args: unknown[]) => unknown

function isFunction<T extends AnyFunction>(value: unknown): value is T {
  return typeof value === 'function'
}
```

**Comment by WoodyWoodsta** [comment]
This cropped up for me today, after updating to v4.6.x from v4.5.5. I'd love to know what exactly changed between these two versions that broke my code. From the above, nobody has really figured out what the root cause might be. If the difference is so simply described (and by that I do not mean so simple a problem) in https://github.com/microsoft/TypeScript/issues/37663#issuecomment-605846042, what's the hold up on a triage here?

**Comment by RyanCavanaugh** [comment]
@WoodyWoodsta this bug is fully triaged. What do you mean?

**Comment by WoodyWoodsta** [comment]
@RyanCavanaugh Might be a difference in our understanding on what "triage" means :) Not sure how the backlog milestone is treated for this project but given this was identified and added to that backlog two years ago, I was wondering if there was anything holding it back given it has now clearly found its way into stable release.

**Comment by keyboard3** [comment]
i think maybe typeof == 'function' canot exclude unknow type?

type Initializer<T> = T | (() => T)
function correct<T>(arg: Initializer<T>) {
    return typeof arg === 'object' && typeof arg == "function" ? arg() : arg // no error
}

type NoUnknown = number|string|never;
function correct2<T extends NoUnknown>(arg: Initializer<T>) {
    return typeof arg === 'function' ? arg() : arg // no error
}

type withUnknown = number|string|never|unknown;
function correct3<T extends withUnknown>(arg: Initializer<T>) {
    return typeof arg === 'function' ? arg() : arg // same error
}

**Comment by keyboard3** [comment]
i think maybe typeof == 'function' canot exclude unknow type?
`
type Initializer = T | (() => T)
function correct(arg: Initializer) {
return typeof arg === 'object' && typeof arg == "function" ? arg() : arg // no error
}

type NoUnknown = number|string|never;
function correct2(arg: Initializer) {
return typeof arg === 'function' ? arg() : arg // no error
}

type withUnknown = number|string|never|unknown;
function correct3(arg: Initializer) {
return typeof arg === 'function' ? arg() : arg // same error
}`

**Comment by keyboard3** [comment]
i think maybe `typeof == 'function'` cannot exclude unknown type?
```
type Initializer = T | (() => T)
function correct(arg: Initializer) {
return typeof arg === 'object' && typeof arg == "function" ? arg() : arg // no error
}

type NoUnknown = number|string|never;
function correct2(arg: Initializer) {
return typeof arg === 'function' ? arg() : arg // no error
}

type withUnknown = number|string|never|unknown;
function correct3(arg: Initializer) {
return typeof arg === 'function' ? arg() : arg // same error
}
```

**Comment by mnpenner** [comment]
I gave up and just casted it.

```ts
// global.d.ts
type Fn<TArgs extends ReadonlyArray<any>=unknown[], TRet=unknown> = (...args: TArgs[]) => TRet
type AnyFn = Fn<any[],any>

// resolvable.ts
export type Resolvable<T = any, TArgs extends ReadonlyArray<any> = []> = T | ((...args: TArgs) => T)

export type Resolved<T> = T extends Fn ? ReturnType<T> : T

export function resolveValue<T, TArgs extends ReadonlyArray<any>>(val: Resolvable<T, TArgs>, ...args: TArgs): T {
    return typeof val === 'function' ? (val as AnyFn)(...args) : val
}
```

**Comment by pspdragon** [comment]
I don't think this is a bug. You use `typeof arg` to identify whether the arg is a function, but from the perspective of the TS compiler, `arg` may not be a function as you declared `T | (() => T)`. The TS compiler doesn't know what your code does, therefore it is not safe to call `arg()`

**Comment by EltonLobo07** [comment]
I think a safer approach would be to let the developer decide what is or  isn't type `T` using a function that uses a type predicate
```ts
function someFuncName<T>(arg: T | (() => T), isTypeT: (val: unknown) => val is T): T {
    if (isTypeT(arg)) {
        return arg;
    }
    return arg();
}
```

**Comment by ITenthusiasm** [comment]
Got more workarounds! lol. Like @EltonLobo07 mentioned, I'm using a predicate to get the job done. But for my code base, the following is sufficient for me to work around the issue.

```ts
type MyUnion<T> = T | ((arg: string) => T);

function doMyLogic<S>(arg: MyUnion<S>): void {
  const stringArg = "my-string";

  // @ts-expect-error -- Weird TypeScript Bug
  const value = typeof arg === "function" ? arg(stringArg) : arg;
  const betterValue = isFunction(arg) ? arg(stringArg) : arg; // Success
}

function isFunction(value: unknown): value is (arg: string) => unknown {
  return typeof value === "function";
}
```

This is obviously not reliable for broad usage. But since I needed a quick TS fix and the impact was minimal, I reached for this. Less verbosity. The narrowed type in the `doMyLogic` function uses the correct return value (rather than `unknown`).

**Comment by AnorZaken** [comment]
This works fine? Or maybe I missed the point...?
```ts
export type ValueTFn<T = any> = T extends ((...args: unknown[]) => unknown) ? never : T | (() => T);
export namespace ValueTFn {
  export function get<T = any>(valueOrFn: ValueTFn<T>): T {
    return (valueOrFn && typeof valueOrFn === "function") ? valueOrFn() : valueOrFn;
  }
}
```

Compiles, runs, and returns what I expect from:
```ts
let a1 = 1;
let a2 = () => 2;
let a3 = (foo: any) => foo;
let a4 = null;
let a5 = "Hello!";
let a6 = () => a1;

console.log(ValueTFn.get(a1)); // 1
console.log(ValueTFn.get(a2)); // 2
console.log(ValueTFn.get(a3)); // undefined
console.log(ValueTFn.get(a4)); // null
console.log(ValueTFn.get(a5)); // "Hello!"
a1 = 42;
console.log(ValueTFn.get(a6)); // 42
```

[Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAaghgGwK4QCoDEB2AeVUC8UcmIAfAVHhAB7ASYAmAzlABSsB0XcATgOZMAXFCSYA1pgD2Ad0wBtALoBKAuVESZmFQH4omCADcIPKMLwAfNqxX5yqJQG4AUJjgBbCEzBwAxtHjIaFhQAN5OUFA0YJI8wFAAZqI+wACWkphQfBDAuBTEZKwGiCgA8jxYwgEoGDiopEpmoeERUDzZSDwZhcUQZcEAZP1QoJCS8VBFgX0Z+LNQAESJmMlpmPM6Ez3T1qabU+WYzhEAvk6nTgjZRACMFNfOl3FwAEwUO7ZQzw9XcADMb-FJJJhPkbORAZJvk8ACwUTBIBAIKFEACsFHmAAkIIjJABCebIuAANjeYJuzicPnSTEklw4CEkfFYVSCmA4WWArDg1yUjkp1NpEHpjOZPRq7OyXOevOcVMwNLpDKZLPFHK5vxl-PlguFyrFWAlnLg0M1coVQqVosCqslcBRmu5FGhXy15t1VuqBrVxJlQA)

**Comment by Irrelon** [comment]
> I don't think this is a bug. You use `typeof arg` to identify whether the arg is a function, but from the perspective of the TS compiler, `arg` may not be a function as you declared `T | (() => T)`. The TS compiler doesn't know what your code does, therefore it is not safe to call `arg()`

It knows there are only two types, type `T` or type `() => T`. `T` can be anything including a function, but the other type IS definitely a function. If you check the type with a `typeof arg === 'function'` then `arg` must be callable. It's definitely a bug given the `arg` is callable because it is a function. It could be considered that the error message is incorrect and that the message should be something like "Type T could be a function that does not match the signature () => T", but that is against the intent of the developer at that point, because we don't have runtime checking of specific arguments etc. How would be go about describing that T can be anything, including a function, but that if it matches a specific call signature and return value, then it should execute some other code.

`
function (arg: T | () => T): T {
    if (typeof arg === "function") { // This should be enough to indicate that arg() is callable
        return arg();
    }

    return arg;
}
`

**Comment by jeremy-code** [comment]
Sorry for commenting on an old thread, but TS 5.5, there is inferred type guards, making the following code valid:

```typescript
const isFunction = (x: unknown) => typeof x === "function";

export const correct = <T,>(initialValue: T | (() => T)) => {
  return isFunction(initialValue) ? initialValue() : initialValue;
};
```

where initialValue is correctly inferred as `() => T` if it is a function, and `T` otherwise. IMO, this is a little better than writing it out the output as `T is Function` since it would sync with the function implementation.

**Comment by WPMGPRoSToTeMa** [comment]
Conditional types can be used to eliminate passing `Function`s with a signature different from `() => R`:
```typescript
function f<T>(v: T extends Function ? (T extends () => any ? T : never) : T) {
    return typeof v === "function" ? v() : v;
}
```
Unfortunately this leads to `f` having `any` return type. We could apply the following return type to the `f`:
```typescript
T extends Function ? (T extends () => infer R ? R : never) : T
```
Though it emits an error ([works in v.5.7.3](https://www.typescriptlang.org/play/?target=99&jsx=0&ts=5.7.3#code/GYVwdgxgLglg9mABMAPAFQHwAoBuAuRNRAUwA8piwATAZ0QDFxp4kB+RLIsi6urASkQBeDIgCGYAJ6J2RAmGI5iAJ0EE0-AFCIdidSXKVaDJrAQyOXQ7w6CRiGGGArEAJQvv5ilWsKIA3tq6ysRQIMpIUJIADsRwwIg4wkJCiABEoJBmYGkWOAJ6iQDcmgC+mprAWBlwcGn8JVUF9mkARmLK9Y1YzaIALABMDZVY-qXDTaQENFDKjgDmdqKkw5pRsYgA4sIcU4gzc2CLwqIHCyUQCDOI8wTbqaQniKQAdFBwAKrRscoAwmI0YgCRooTbYRZFIA), but not later). It can be worked around via type assertions, not ideal but it works. The final solution:
```typescript
function f<T>(v: T extends Function ? (T extends () => any ? T : never) : T) {
    return (typeof v === "function" ? v() : v) as
        T extends Function ? (T extends () => infer R ? R : never) : T;
}
```
A few examples:
```typescript
f("foo"); // => "foo"
f(() => "bar"); // => "bar"
f(() => 42); // => 42
f({}); // => {}
f((x: string) => x); // error
```
Moreover it even prevents the following case (which was mentioned in [this SO thread](https://stackoverflow.com/questions/76201173/how-to-safely-call-a-function-in-typescript-that-was-passed-as-a-parameter-of-t)):
```typescript
type G = (x: string) => string;
const g: G = x => x.toUpperCase();
f<G>(g); // error
```
[Playground](https://www.typescriptlang.org/play/?target=99&jsx=0#code/GYVwdgxgLglg9mABMAPAFQHwAoBuAuRNRAUwA8piwATAZ0QDFxp4kB+RLIsi6urASkQBeDIgCGYAJ6J2RAmGI5iAJ0EE0ggN4AoRHsTLiUEMqRYokgA7E4wRDmFChiAEShIsBC5n2BiAjiCYjS6+mFc5JS0DEyebBwRPNF+IogwYMAqiABKPrnyiipqhADc2gC+2trAWG5wcC78ZTUpoi4ARmLKjc1YrYgALABMTdVYmuWjLaQENFDK6QDmgqmko9oW1ogA4sIcM4hzC2DLwqJHS2UQCHOIiwS7zqRniKQAdFBwAKqW1soAwsFiAJmihtthliUgA)

--------------------------------------------
The solution above is actually a variation of [this one posted on SO](https://stackoverflow.com/questions/76201173/how-to-safely-call-a-function-in-typescript-that-was-passed-as-a-parameter-of-t#answer-76206997). It's even shorter:
```typescript
function f<T>(v: T extends Function ? never : T | (() => T)) {
    return (typeof v === "function" ? v() : v) as T;
}
```
[Playground](https://www.typescriptlang.org/play/?target=99&jsx=0#code/GYVwdgxgLglg9mABMAPAFQHwAoBuAuRNRAUwA8piwATAZ0QDFxp4kB+RMYnYgJ0QKIAfRFiwBKRAF4MhMRIDeAKEQrEPYlBA8kWKAE8ADsTjBEOKZMmIARKEiwE1xOxzj+ZiQEM6aANyKAX0VFYCxbODhrMX9Qt2kbACNPHiiY0Ql4gBYAJmiQrHkAvNjSAhooHhgwAHMMmVI8xX0jRABxKRFSxHLKmrruiqrq-wgEcsRqgnarUil6gDooOABVAyMeAGFvYnEYlFbsWt8gA)

In v5.7.3 and earlier it can be written as:
```typescript
function f<T>(v: T extends Function ? never : T | (() => T)): T {
    return typeof v === "function" ? v() : v;
}
```
[Playground](https://www.typescriptlang.org/play/?target=99&jsx=0&ts=5.7.3#code/GYVwdgxgLglg9mABMAPAFQHwAoBuAuRNRAUwA8piwATAZ0QDFxp4kB+RMYnYgJ0QKIAfRFiwBKRAF4MhMWIGIA3gChEaxD2JQQPJFACeAB2JxgiHFMmTEAIlCRYCG4nY5x-cwG5lAX2XLgLDs4OBsxb0D3aVsAIwBDHjCI0QlogBYAJnCArEUfbMjSAhooHhgwAHNUmVJs5QNjRABxKREixBKyyuqO0vKK7wgEEsQKghbrUikagDooOABVQ2MeAGE4mmJxCJQm7CrPIA)
